pkg/networking/security_group_resolver.go

```diff
@@ -102,3 +109,102 @@ func (r *defaultSecurityGroupResolver) splitIntoSgNameAndIDs(sgNameOrIDs []strin
+		if err != nil {
+			return nil, err
+		}
+		if len(chosenSGs) != len(selector.IDs) {
+			return nil, fmt.Errorf("couldn't find all security groups: IDs: %v, found: %v", selector.IDs, len(chosenSGs))
+		}
+	} else {
+		req := &ec2sdk.DescribeSecurityGroupsInput{
+			Filters: []*ec2sdk.Filter{
+				{
+					Name:   awssdk.String("vpc-id"),
+					Values: awssdk.StringSlice([]string{r.vpcID}),
+				},
+			},
+		}
+		for key, values := range selector.Tags {
+			req.Filters = append(req.Filters, &ec2sdk.Filter{
+				Name:   awssdk.String("tag:" + key),
+				Values: awssdk.StringSlice(values),
+			})
+		}
+
+		allSGs, err := r.ec2Client.DescribeSecurityGroupsAsList(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		var filteredSGs []*ec2sdk.SecurityGroup
+		for _, sg := range allSGs {
+			if r.checkSecurityGroupIsNotTaggedForOtherClusters(sg) {
+				filteredSGs = append(filteredSGs, sg)
+			}
+		}
+		for _, sg := range filteredSGs {
+			if r.checkSecurityGroupHasClusterTag(sg) {
+				chosenSGs = append(chosenSGs, sg)
+			}
+		}
+		if len(chosenSGs) == 0 {
+			chosenSGs = filteredSGs
+		}
+	}
+	if len(chosenSGs) == 0 {
+		return nil, errors.New("unable to resolve at least one security group")
+	}
+	resolvedSGIDs := make([]string, 0, len(chosenSGs))
+	for _, sg := range chosenSGs {
+		resolvedSGIDs = append(resolvedSGIDs, awssdk.StringValue(sg.GroupId))
+	}
+	sort.Strings(resolvedSGIDs)
+	return resolvedSGIDs, nil
+}
+
+// checkSecurityGroupHasClusterTag checks if the subnet is tagged for the current cluster
+func (r *defaultSecurityGroupResolver) checkSecurityGroupHasClusterTag(securityGroup *ec2sdk.SecurityGroup) bool {
+	clusterResourceTagKey := fmt.Sprintf("kubernetes.io/cluster/%s", r.clusterName)
+	for _, tag := range securityGroup.Tags {
+		if clusterResourceTagKey == awssdk.StringValue(tag.Key) {
+			return true
+		}
+	}
+	return false
+}
+
+// checkSecurityGroupIsNotTaggedForOtherClusters checks whether the security group is tagged for the current cluster
+// or it doesn't contain the cluster tag at all. If the security group contains a tag for other clusters, then
+// this check returns false so that the security group is not used for the load balancer.
+func (r *defaultSecurityGroupResolver) checkSecurityGroupIsNotTaggedForOtherClusters(sg *ec2sdk.SecurityGroup) bool {
+	clusterResourceTagPrefix := "kubernetes.io/cluster"
+	clusterResourceTagKey := fmt.Sprintf("kubernetes.io/cluster/%s", r.clusterName)
+	hasClusterResourceTagPrefix := false
+	for _, tag := range sg.Tags {
+		tagKey := awssdk.StringValue(tag.Key)
+		if tagKey == clusterResourceTagKey {
+			return true
+		}
+		if strings.HasPrefix(tagKey, clusterResourceTagPrefix) {
+			// If the cluster tag is for a different cluster, keep track of it and exclude
+			// the security group if no matching tag found for the current cluster.
+			hasClusterResourceTagPrefix = true
+		}
+	}
+	if hasClusterResourceTagPrefix {
```

Comment:
Maybe it's better to add a log here just to indicate we skip some SG because it's tagged for another cluster `cluster-xxx`

Reply:
I incorporated code corresponding to #3292